### PR TITLE
[Core][Distributed] refactor pynccl to hold multiple communicators

### PR DESCRIPTION
**Pull request by youkaichao**
Currently pynccl is bound to the module level instance. And we are using it for just tensor parallel group.

After this refactor, we can create as many pynccl communicator instances as we want, e.g. a new pynccl communicator for pipeline parallel group.

This is an ongoing effort to support pipeline parallel https://github.com/vllm-project/vllm/pull/4412 .

### CODE REVIEW

**Comment by youkaichao** [comment]
Major change:

Prior to this PR:
1. we use `pynccl_utils.comm is None` to indicate pynccl is not available
2. we use `parallel_state._ENABLE_PYNCCL_FOR_ALL_REDUCE == False` to control if `pynccl_utils.comm` will be used in allreduce, which is `False` by default
3. we use `model_runner._maybe_pynccl()` context manager to temporarily enable pynccl for allreduce if pynccl is available but custom allreduce is not available.

This PR takes an OOP approach, absorb these pynccl related state and knobs into `NCCLCommunicator`, and detach it from module level instance:
1. `parallel_state._TP_PYNCCL_COMMUNICATOR` is always a `NCCLCommunicator` instance after `initialize_model_parallel` is called. Later we can also add one `parallel_state._PP_PYNCCL_COMMUNICATOR` for pipeline parallel.
2. we use `NCCLCommunicator.available == False` to indicate pynccl is not available
3. we use `NCCLCommunicator.disabled == True` to indicate if this communicator will be used in allreduce, which is `False` by default
4. we use `NCCLCommunicator.enable()` context manager to temporarily enable pynccl for allreduce, and use `communication_op.use_pynccl_allreduce()` context manager to call `NCCLCommunicator.enable()` if pynccl is available but custom allreduce is not available.

The effect should be equivalent with code before refactor. This PR is basically a cleanup PR to make pynccl related code clean and extensible to more parallel strategies.

Next steps:

- [ ] export more functions (e.g. send/recv) from NCCL to support pp
- [ ] enable custom allreduce for multiple tp groups (e.g. when we have 2 node and 8GPUs in each node, intra-node communication is NVSwitch and thus is fully connected by NVLink, inter-node communication is Infiniband. the world size ==16, and currently custom allreduce will be disabled in this case because 16 GPUs are not fully connected. However, when we divide the world size into 2 tensor parallel groups, each node corresponds to one tp group, custom allreduce can be enabled inside each node.)

File: vllm/distributed/parallel_state.py [hunk 1]
```diff
@@ -3,21 +3,22 @@
...
 
 # Tensor model parallel group that the current rank belongs to.
-_TP_DEVICE_GROUP = None
-_TP_CPU_GROUP = None
+_TP_DEVICE_GROUP: Optional[ProcessGroup] = None
+_TP_CPU_GROUP: Optional[ProcessGroup] = None
+_TP_PYNCCL_COMMUNICATOR = None
 # Pipeline model parallel group that the current rank belongs to.
-_PIPELINE_MODEL_PARALLEL_GROUP = None
+_PIPELINE_MODEL_PARALLEL_GROUP: Optional[ProcessGroup] = None
```
Comment (WoosukKwon) [inline vllm/distributed/parallel_state.py:21]: nit: Let's use `PP` instead of `PIPELINE` for consistency with `TP`?

File: vllm/distributed/parallel_state.py [hunk 2]
```diff
@@ -80,10 +81,20 @@ def init_distributed_environment(
...
+                local_rank = envs.LOCAL_RANK
+            else:
+                local_rank = rank
         global _LOCAL_RANK
         _LOCAL_RANK = local_rank
+        # A small all_reduce for warmup.
+        data = torch.zeros(1)
+        if torch.cuda.is_available():
+            data = data.to(device=f"cuda:{local_rank}")
+        torch.distributed.all_reduce(data)
```
Comment (WoosukKwon) [inline vllm/distributed/parallel_state.py:97]: I feel warmup should not be a part of this method?

File: vllm/distributed/parallel_state.py [hunk 3]
```diff
@@ -133,25 +144,32 @@ def initialize_model_parallel(
...
+        ranks = list(
+            range(i * tensor_model_parallel_size,
+                  (i + 1) * tensor_model_parallel_size))
         group = torch.distributed.new_group(ranks, backend=backend)
         cpu_group = torch.distributed.new_group(ranks, backend="gloo")
         if rank in ranks:
             _TP_DEVICE_GROUP = group
             _TP_CPU_GROUP = cpu_group
 
+    from vllm.distributed.device_communicators.pynccl import NCCLCommunicator
```
Comment (WoosukKwon) [inline vllm/distributed/parallel_state.py:160]: Again, why do we need this lazy import?

**Review by WoosukKwon — CHANGES_REQUESTED** [review]
@youkaichao Thanks a lot for the PR! The refactoring makes sense.

Left some comments on code style and possible errors in the PR. Please check my review.

File: vllm/distributed/parallel_state.py (hunk 3 above)
Comment (youkaichao) [reply vllm/distributed/parallel_state.py:160]: lazy import is required here to avoid circular import. `vllm.distributed.device_communicators.pynccl` will try to import `vllm/distributed/parallel_state.py`.

File: vllm/distributed/communication_op.py [hunk 4]
```diff
@@ -23,18 +36,19 @@ def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
     TLDR: always assume this function modifies its input, but use the return
     value as the output.
     """
-    from vllm.distributed.device_communicators import pynccl_utils
     from vllm.distributed.device_communicators.custom_all_reduce import (
         custom_all_reduce)
+    from vllm.distributed.parallel_state import _TP_PYNCCL_COMMUNICATOR
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/distributed/communication_op.py:41]: > Why don't we import these at the top of the module?

Because this variable is `None` when the module gets initialized. Import these at the top of the module will get `None`. It will not be updated as the variable in the `vllm.distributed.parallel_state`.

> the variable is used outside the module where it's defined

I actually want to merge `parallel_state` and `communication_op` . `communication_op` needs to access many private variable in `parallel_state` . That's another story though.

File: tests/distributed/test_pynccl.py [hunk 5]
```diff
@@ -41,6 +40,10 @@ def worker_fn_wrapper(fn):
     # and update the environment variables in the function
     def wrapped_fn(env):
         update_environment_variables(env)
+        import os
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply tests/distributed/test_pynccl.py]: fixed in [0516956](https://github.com/vllm-project/vllm/pull/4591/commits/0516956ce52771a5d1495c171463d45698d3a2b5)

File: tests/distributed/test_pynccl.py [hunk 6]
```diff
@@ -51,7 +54,8 @@ def wrapped_fn(env):
 def worker_fn():
     comm = NCCLCommunicator()
     tensor = torch.ones(16, 1024, 1024, dtype=torch.float32).cuda(comm.rank)
-    comm.all_reduce(tensor)
+    with comm.enable():
+        comm.all_reduce(tensor)
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply tests/distributed/test_pynccl.py]: fixed in [9f63bf8](https://github.com/vllm-project/vllm/pull/4591/commits/9f63bf8a773f847837c2e2c7992a19cfcb1c3fef)

File: tests/distributed/test_pynccl.py [hunk 7]
```diff
@@ -95,12 +100,9 @@ def test_pynccl_multiple_tp():
 @worker_fn_wrapper
 def multiple_tp_with_vllm_worker_fn():
     device = torch.device(f"cuda:{torch.distributed.get_rank()}")
-    torch.cuda.set_device(torch.distributed.get_rank())
     ensure_model_parallel_initialized(2, 2)
-    pynccl_utils.init_process_group(
-        group=get_tensor_model_parallel_cpu_group())
     tensor = torch.ones(16, 1024, 1024, dtype=torch.float32, device=device)
-    with with_pynccl_for_all_reduce():
+    with use_pynccl_allreduce():
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply tests/distributed/test_pynccl.py]: fixed in [e9aa766](https://github.com/vllm-project/vllm/pull/4591/commits/e9aa766c278166db7d141117c489d498a54af29d)

File: tests/distributed/test_pynccl.py [hunk 8]
```diff
@@ -129,7 +131,7 @@ def worker_fn_with_cudagraph():
         # run something in the default stream to initialize torch engine
         a = torch.ones((4, 4), device=f'cuda:{comm.rank}')
         torch.cuda.synchronize()
-        with torch.cuda.graph(graph, stream=comm.stream):
+        with torch.cuda.graph(graph, stream=comm.stream), comm.enable():
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply tests/distributed/test_pynccl.py]: > Why do we need comm.enable here?

Because pynccl is disabled in eager mode (the default mode).

> Also, IIRC, python 3.8 does not support multiple with statements in a single line?

The code works fine in python 3.8:

```python
from contextlib import nullcontext

with nullcontext(), nullcontext():
    print("Hello, world!")
```

File: vllm/distributed/communication_op.py [hunk 9]
```diff
@@ -23,18 +36,19 @@ def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
...
     # Bypass the function if we are using only 1 GPU.
     if get_tensor_model_parallel_world_size() == 1:
         return input_
     out = custom_all_reduce(input_)
     if out is not None:
         return out
-    if is_pynccl_enabled_for_all_reduce():
-        pynccl_utils.all_reduce(input_)
+    if _TP_PYNCCL_COMMUNICATOR is not None and \
+        not _TP_PYNCCL_COMMUNICATOR.disabled:
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/distributed/communication_op.py]: fixed in [68e448c](https://github.com/vllm-project/vllm/pull/4591/commits/68e448c8039c3d6ee9d939ebd88d6e90d1a0c81f).

File: vllm/distributed/device_communicators/pynccl.py [hunk 10]
```diff
@@ -261,15 +91,25 @@ def __init__(
...
+        # to use it, we have to use under `with obj.enable()`, usually
+        # when we are using CUDA graph.
+        self.disabled = True
+
     def all_reduce(self,
                    tensor: torch.Tensor,
                    op: ReduceOp = ReduceOp.SUM,
                    stream=None):
+        if self.disabled:
+            return
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/distributed/device_communicators/pynccl.py:112]: No, the semantic is the same as custom allreduce. If allreduce fails, just return None.

File: vllm/distributed/parallel_state.py (hunk 2 above)
Comment (youkaichao) [reply vllm/distributed/parallel_state.py:97]: which method do you think is better?

**Comment by youkaichao** [comment]
@WoosukKwon one change after our discussion:

[c1b1cdb](https://github.com/vllm-project/vllm/pull/4591/commits/c1b1cdbce0934fc5d366cad103a9a7877d86daea) change `with pynccl_comm.enable()` to `with pynccl_comm.change_state(enable=True)` . I think this makes more sense.

File: vllm/distributed/parallel_state.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/distributed/parallel_state.py]: If that's the case, I think it means it's a bad design tbh. We should use lazy import only to avoid the unnecessary imports, but to avoid circular imports. Otherwise, the code will be too complicated.

File: vllm/distributed/communication_op.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/distributed/communication_op.py:47]: Why don't we use a get method then?

[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/distributed/communication_op.py:47]: a get method is doable. however, if the majority callers of these getter method come from `communication_op`, it's better to simply merge two files, imo.

[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/distributed/communication_op.py:47]: Maybe yes. But if we do not merge the files in this PR, I think using the getter method is a good idea. The current code has two problems: 1) It must be lazily imported for the reason you mentioned above, and 2) the variable name starts with `_`. I believe using the getter method is a simple solution to the two problems.

[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/distributed/communication_op.py:47]: I don't want this PR to be super complicated. I will merge the files in the next days, let's don't add new user-facing getter functions just for several days and remove them later 🤣

[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/distributed/communication_op.py:47]: Honestly, I think in this case we should add the getter function even if there's a change that it can be deleted just after several days. This is because we don't have a concrete timeline for the next PR you are thinking of. It can be delayed or de-prioritized any way. I believe adding the getter function doesn't hurt us.

[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/distributed/communication_op.py]: Added in [70a7e26](https://github.com/vllm-project/vllm/pull/4591/commits/70a7e2679b477e02c6559c4c47f9368d7c728ad3) .

**Review by WoosukKwon — APPROVED** [review]
LGTM. Thanks for addressing my review! Looking forward to the planned refactoring!
